File: jun/2024-07-28-subsets.ts
function subsets(nums: number[]): number[][] {
    const result: number[][] = []
    const { length } = nums
    function helper(i, subset) {
        if (i >= length) {
            result.push(subset)
            return
        }
        helper(i + 1, [...subset, nums[i]])
        helper(i + 1, [...subset])
    }
    helper(0, [])
    return result
}

File: Changwhi/2024-07-28-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []

        subset = []
        def backtracking(index):
            if index >= len(nums):
                result.append(subset[:])
                return

            subset.append(nums[index])
            backtracking(index+1)

            subset.pop()
            backtracking(index+1)
        backtracking(0)
        return result